### ISSUE TITLE
Cancelling an async SqlClient operation throws SqlException, not TaskCanceledException

### ISSUE THREAD

**Issue by roji** [dotnet/corefx#9245]
When using a cancellation token to cancel an SqlClient async operation, an SqlException is thrown rather than TaskCanceledException. This seems like problematic behavior, here's some analysis.

On the problematic side, standard async APIs generally throw TaskCanceledException, and SqlClient is doing something different and therefore unexpected. Also, as a result, canceled SqlClient tasks don't have status Canceled, they're Faulted instead. This makes differentiating cancellations from real exceptions more difficult, and the problem is compounded if the SqlClient call is just part of a larger async operation (which again, will be faulted instead of canceled unless some manual specific handling is involved).

The argument for the current behavior seems to be that all of SqlClient's server and network errors are raised as SqlException, making it easy to catch all database errors under a single exception type. On the other hand, it seems that cancellation isn't really an error condition as it's been requested by the user and is expected to occur. The TaskCanceledException is simply a mechanism for conveying the cancellation up the stack.

For the record, this was originally reported by @mikkeljohnsen for Npgsql in https://github.com/npgsql/npgsql/issues/1146. Whatever is decided here for SqlClient will also be adopted for Npgsql (which currently behaves like SqlClient in this respect).

**Comment by GSPP** [comment]
If this is changed the exception should allow to tell the difference between:

1. Timeout
2. `SqlCommand.Cancel`
3. Some other external cancellation arrived through a `CancellationToken`

My fear is that it becomes hard to analyze error logs. Also, sometimes programmatic action is required. Maybe you don't want to send an error mail to the ops team for explicit cancellations but you do want to send one for timeouts.

`TaskCanceledException` is not exactly the nicest exception possible and it's quite nasty already with `HttpClient`.

**Comment by roji** [comment]
@GSPP, I think distinguishing between timeouts and cancellations would be easy in any case - TaskCanceledException would only be thrown for cancellations, not timeouts. SqlException already provides the means to programmatically understand exactly what happened. The question here is only whether cancellations should be be raised as SqlException or as TaskCanceledException.

**Comment by GSPP** [comment]
Good point.

**Comment by bgrainger** [comment]
When developing [MySqlConnector](https://github.com/mysql-net/MySqlConnector), I didn't check the SqlClient behaviour first, but just assumed that throwing `OperationCanceledException` (and transitioning the `Task` returned from `ExecuteXAsync` to the `Canceled` state) was the right thing to do. (At the moment, I plan to keep this behaviour instead of changing it to match SqlClient and Npgsql.)

Throwing:
https://github.com/mysql-net/MySqlConnector/blob/a06bf08763fa97f3e4c7c0157a0e54d9d5b8bccf/src/MySqlConnector/MySqlClient/CommandExecutors/TextCommandExecutor.cs#L64-L67

Example test:
https://github.com/mysql-net/MySqlConnector/blob/a06bf08763fa97f3e4c7c0157a0e54d9d5b8bccf/tests/SideBySide/CancelTests.cs#L186

**Comment by mikkeljohnsen** [comment]
@bgrainger Yes it is the logical thing to do. The SqlClient is doing it wrong and should change it ASAP. 

I think Npgsql should follow your lead and do the right thing.

**Comment by roji** [comment]
As this hasn't even received a comment in such a long time, I'll go ahead and change Npgsql's behavior for 3.3.

I really do hope Microsoft improves their attitude towards database access and ADO.NET, it's quite frustrating for things to receive zero attention at all - Entity Framework is important but so is raw SQL access (@divega maybe you can help push again)

**Comment by JackLiang5** [comment]
Ref to _Cancellation in Managed Threads_ ( https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads?view=netframework-4.7.2 )

Cancellations should be raised as OperationCanceledException or its derived class (e.g. TaskCanceledException). Usually, exception handler (including try/catch) will process OperationCanceledException specially. There're some examples in referred page. Raising other exception like SqlException will break these examples.

**Comment by divega** [comment]
As recently [announced in the .NET Blog](https://devblogs.microsoft.com/dotnet/introducing-the-new-microsoftdatasqlclient/), focus on new SqlClient features an improvements is moving to the new [Microsoft.Data.SqlClient package](https://www.nuget.org/packages/Microsoft.Data.SqlClient/). For this reason, we are moving this issue to the new repo at https://github.com/dotnet/SqlClient. We will still use https://github.com/dotnet/corefx to track issues on other providers like System.Data.Odbc and System.Data.OleDB, and general ADO.NET and .NET data access issues.
